### DISCUSSION TITLE
state resets when transitioning route(page)

### DISCUSSION THREAD

**Discussion by mohdsauood** [vercel/next.js#17377] [Help] [answered]
i recreated the problem in this [CodeSandbox][1]

https://codesandbox.io/s/inspiring-pond-mp134?file=/context/cars-context.js

i have two routes(pages)
- `/` 
- `/car/info`

and two components
- Header
- CarInput

i used context to maintain the state and pass it to both the pages.

when im in route `/car/info` and i change the state using input , the updated state gets displayed in Header component of the same route(page).

But when i go to `/` route(page) the state is not updated .i inspected using react dev tools and the context doesnt have the updated values . 

Even when i go back to `/car/info` , the updated state is not there too. As soon as i transition to `/` from `/car/info` the state goes back to intialState . 

Steps to recreate the same on sandbox

 1. click on "go to add car"
 2.  enter a car name in the input and click add
 3. click on "go to home"
 
 Now you can see the car you added is not present there .

Context Usage Pattern reference : https://kentcdodds.com/blog/how-to-use-react-context-effectively

  [1]:https://codesandbox.io/s/inspiring-pond-mp134?file=/context/cars-context.js

**Comment by lfades** [comment, marked as answer]
You need to put your context provider in a place where it doesn't get unmounted when you do a page transition. For that you can use [`pages/_app.js`](https://nextjs.org/docs/advanced-features/custom-app) and wrap your page with the provider.

**Reply by mohdsauood** [reply]
thanks for the reply ! Much appreciated

**Reply by jbbae** [reply]
Is there a more elegant way than this solution?

**Reply by arshad-yaseen** [reply]
@lfades We need more elegant way!

**Comment by GaddMaster** [comment]
So what on earth are we supposed to do with the NextJS page changes and the data is fetched from getServerSideProps ?

I assume the best thing to do is stick Provider in _app so it's only mounted once and not each page change, and I am assuming we may need use the useEffect hook in _app so monitor changes and call dispatch on page changes if required ?

Daniel

**Comment by JacobDel** [comment]
The installation tutorial (https://nextjs.org/docs/getting-started/installation) says "If you prefer to use the Pages Router **instead of the App Router**, you can create a pages/ directory at the root of your project."
So if _app.tsx isn't an option, what options are left?....

**Comment by miltonrincon** [comment]
I had the same issue on Nextjs 14.2.18, for me that wasnt an option, so i resolved it, not using next/navigation, I changed the window href

* router.push(route) // I removed this

* window.location.href = route // To use this

Problem resolved, for my case, with app router.
